router.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package ringpop
+
+import (
+	"sync"
+
+	"github.com/uber/ringpop-go/swim"
+)
+
+// Replication constants
+type ReplicationFactor int
+
+const (
+	ALL ReplicationFactor = -1 * (iota + 1)
+	// TODO implement quorum
+	// QUORUM
+)
+
+type router struct {
+	ringpop RingpopInterface
+	factory ClientFactory
+
+	rw          sync.RWMutex
+	clientCache map[string]interface{}
+}
+
+type Router interface {
+	GetClients(key string, n ReplicationFactor, fanout ReplicationFactor) []interface{}
+}
+
+type ClientFactory interface {
+	MakeClient(hostport string) interface{}
+}
+
+func NewRouter(rp RingpopInterface, f ClientFactory) Router {
+	r := &router{
+		ringpop:     rp,
+		factory:     f,
+		clientCache: make(map[string]interface{}),
+	}
+	rp.RegisterListener(r)
+	return r
+}
+
+func (r *router) HandleEvent(event interface{}) {
```

Comment:
This implements the `ringpop.EventListener` interface, so yes this needs to be exported. It is only exported on the `router` type -which in its self is not exported- so it is not accessible on the `Router` interface returned from the constructor function.